### PR TITLE
Fix client logging with cancelled content (#1735).

### PR DESCRIPTION
**Pull request by dmitrievanthony**
**Subsystem**
Client, Logging

**Motivation**
https://github.com/ktorio/ktor/issues/1735

### CODE REVIEW

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/observer/ResponseObserver.kt [hunk 1]
```diff
@@ -42,7 +42,7 @@ class ResponseObserver(
         override fun install(feature: ResponseObserver, scope: HttpClient) {
 
             scope.receivePipeline.intercept(HttpReceivePipeline.Before) { response ->
-                val (loggingContent, responseContent) = response.content.split(response)
+                val (loggingContent, responseContent) = response.content.splitOnMainAndDependant(response)
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/features/observer/ResponseObserver.kt]: Could you tell me why we need new method here? It looks like the split should propagate cancellation to all channels

File: ktor-utils/common/src/io/ktor/util/ByteChannels.kt [hunk 2]
```diff
@@ -10,6 +10,37 @@ import kotlinx.coroutines.*
...
+fun ByteReadChannel.splitOnMainAndDependant(coroutineScope: CoroutineScope): Pair<ByteReadChannel, ByteReadChannel> {
+    val first = ByteChannel(autoFlush = true)
+    val second = ByteChannel(autoFlush = true)
+
+    coroutineScope.launch {
+        try {
+            while (!isClosedForRead) {
+                readRemaining(CHUNK_BUFFER_SIZE).use { chunk ->
+                    second.writePacket(chunk.copy())
+                    while (second.availableForRead != 0) { yield() }
```
Comment (e5l) [inline ktor-utils/common/src/io/ktor/util/ByteChannels.kt]: codestyle

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/observer/ResponseObserver.kt (hunk 1 above)
Comment (dmitrievanthony) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/features/observer/ResponseObserver.kt]: Method `split` does split the channel and returns two fully asynchronous channels. As a result, logging can be fully done and completed even before the "main" channel will be canceled.

From my perspective, it's wrong and the logging channel should propagate only the data the main channel already read (it should wait for the main channel). To implement this idea we need a separate `splitOnMainAndDependant` method.

File: ktor-utils/common/src/io/ktor/util/ByteChannels.kt (hunk 2 above)
Comment (dmitrievanthony) [reply ktor-utils/common/src/io/ktor/util/ByteChannels.kt]: Fixed.

**Comment by e5l** [comment]
Fixed in `1.5.0`

**Pull request closed without merging**